### ISSUE TITLE
[RFC]: SM90 Exact Fused DSA Prefill Indexer

### ISSUE THREAD

**Issue by fxfxfxfxfxfxfxfx** [vllm-project/vllm#53563]
### Motivation.

We propose an optional SM90 backend for vLLM's `SparseAttnIndexer` prefill path. It computes FP8 scores and exact TopK in one CUDA kernel without materializing the full score tensor or intermediate candidates in HBM.

The first implementation PR targets only the H64, D=128, topk=2048, single-GPU FP8 prefill path used by DeepSeek V3.2. The existing path remains the default. H32 support will follow in a separate PR.

The current vLLM prefill indexer follows this high-level flow:

```text
gather K -> fp8_fp4_mqa_logits -> materialize logits -> top_k_per_row_prefill
```

At long context lengths, writing and reading the full logits tensor and the extra kernel boundary become significant. In our current H20 matrix, the `H=1M, B=32, Q=512` case has a 508.25 MiB peak score chunk and materializes 64.03 GiB of scores over one indexer call, while the final TopK index output is 128 MiB.

The goal is to replace only the score + TopK stage and eliminate full-score HBM materialization without changing downstream Sparse MLA attention.

#### Scope and integration

The new backend plugs into the existing `sparse_attn_indexer()` prefill branch:

```text
existing Q quantization, KV-cache insert, and chunk metadata
                              |
                              v
                 cp_gather_indexer_k_quant_cache
                              |
                              v
                 SM90 fused score + exact TopK
                              |
                              v
                   existing topk_indices_buffer
```

The first PR is intentionally limited to:

- SM90a, validated separately on H20, H800, and H100;
- the DeepSeek V3.2 FP8 prefill path;
- H64, D=128, and topk=2048;
- single-GPU execution without DCP or PCP;
- INT32 output shaped `[M, 2048]`, with unspecified order.

Decode, KV-cache layout, K gather, and Sparse MLA attention remain unchanged.

The proposed configuration is:

```python
attention_config.indexer_prefill_backend = "default" | "sm90_exact"
```

`default` preserves the existing vLLM implementation. `sm90_exact` explicitly enables the experimental backend. Selecting `sm90_exact` reports a clear initialization error when GPU, dtype, H/D/topk, DCP/PCP, or shared-memory requirements are not met.

Rows shorter than topk use the existing short-sequence path and `-1` sentinel rule. A row exactly equal to topk returns all valid indices without reading Q/K.

DCP is excluded because the current DCP merge consumes both local TopK scores and indices, while this kernel outputs indices only.

#### Kernel design

The current H64 kernel uses a one-query-per-CTA pipeline:

```text
TMA K/K-scale double buffer
        -> 4 WGMMA math warp groups (256 keys/round)
        -> generation-tagged shared-memory score ring
        -> FP32 ordered-bit histogram/radix cutoff
        -> running Top-2048 and final indices
```

The main implementation parameters are:

- 768 threads per CTA;
- four WGMMA math warp groups;
- two TMA stages;
- 12,544 scores per selection group and two ring slots;
- 704 threads participating in the final drain;
- 80 registers per thread;
- 232,448 bytes of dynamic shared memory and 1,024 bytes of static shared memory;
- zero stack and local-memory usage in the production cubin.

Shared-memory usage gives one CTA per SM on H20. H20, H800, and H100 share one SM90a source, while eligibility and performance are validated separately on each device instead of maintaining SKU-specific implementations.

The prototype uses runtime NVCC for iteration speed. The upstream version will retain only the selected production variant and use vLLM's static CUDA build, with no first-request JIT compilation.

#### Exactness contract

`exact` means that the selector computes an exact cutoff for the FP32 scores produced by the fused kernel. For a valid interval, let `c` be its 2048th-largest score. A valid output must:

- contain 2048 unique indices from the valid interval;
- include every index with `score > c`;
- include no index with `score < c`;
- fill the remaining positions from indices with `score == c`.

Output order and the specific choice among cutoff ties are unstable and are not part of the API contract. This matches unordered TopK set semantics.

The score expression is arithmetically equivalent to the current FP8 indexer, but WGMMA and DeepGEMM may use different floating-point reduction orders. Validation must therefore separate selector exactness relative to the fused scores from numerical error and model outputs relative to the latest vLLM backend. This RFC does not claim bitwise equivalence between those score producers.

#### Related work

- **Current vLLM:** materializes full logits in HBM before TopK and serves as the current baseline.
- **[#52696](https://github.com/vllm-project/vllm/pull/52696):** materializes FP16 logits and uses native FP16 TopK, reducing the size of the logits tensor without removing its HBM round trip.
- **[#48726](https://github.com/vllm-project/vllm/pull/48726) and the [LiteTopK paper](https://arxiv.org/abs/2607.11976):** target SM100 with tcgen05 and online candidate filtering, and also include LiteDSA attention work.
- **This proposal:** targets SM90 with TMA/WGMMA, keeps scores on chip, performs exact cutoff selection, and changes only the prefill indexer.

This proposal has a different architecture, selector, and submission scope from #48726, so it should be reviewed as an independent backend rather than appended to that PR.

#### Preliminary results

The following preliminary operator-level results use one NVIDIA H20 with synthetic FP8 inputs and fixed H64, D=128, and topk=2048.

The benchmark matrix contains 665 shapes:

- history KV lengths: 4K, 6K, 8K, 12K, 16K, 24K, 32K, 48K, 64K, 96K, 128K, 192K, 256K, 384K, 512K, 640K, 768K, 896K, and 1M;
- batch sizes: 1, 4, 16, 32, and 128;
- query lengths: 8, 64, 512, 1024, 2048, 4096, and 8192.

The vLLM baseline is vLLM 0.27.1's `fp8_fp4_mqa_logits` with FP32 scores, followed by `top_k_per_row_prefill`. Both paths use the same shapes and mathematical boundaries, but were measured in independent runs. The following speedups are preliminary operator-level comparisons, not paired regression results.

Current results:

- 1.3918981x geometric-mean speedup over vLLM 0.27.1 FP32 DeepGEMM + TopK across all 665 shapes;
- faster than vLLM in 604 of 665 shapes;
- 58 of 58 current correctness cases passed;
- zero sampled DeepGEMM cutoff deficit.

<img width="1532" height="838" alt="Image" src="https://github.com/user-attachments/assets/f94218cb-7fd7-4eff-a8fa-b31b9d03a78d" />

The figure shows four representative shapes covering `B=1/32`, `Q=512/8192`, and 4K--1M history. The 1.3918981x value is calculated from the complete 665-shape matrix, not inferred from these four curves.

These results only justify upstream evaluation. The baseline is still vLLM 0.27.1's FP32 synthetic operator path, and the vLLM and fused-kernel results come from independent runs. Latest-main, H800/H100, real-model, CUDA Graph, and end-to-end TTFT validation are still pending.

#### Compatibility and scope

PR 1 excludes decode, H32, DCP/PCP, FP4/MXFP4, LiteDSA attention, and experimental selector variants. It does not change the default backend or introduce a new KV-cache format.

H32 changes the Q layout, WGMMA epilogue, shared-memory budget, and thread-role balance. It will be submitted separately for GLM-5.1/5.2 after the interface and validation method are established.

#### Test plan

Before requesting implementation review, the PR will provide:

- a cutoff oracle built from a test-only fused-score output, requiring 100% TopK set equality for unique scores and the contract above for ties;
- random and real-model inputs, all-equal and cutoff-dense scores, extreme gates, ragged and unaligned rows, `0/<K/=K/>K`, and multiple ring generations;
- custom-op schema, FakeTensor, output mutation, CUDA Graph, multiple streams, dispatcher error/bypass, and compute-sanitizer coverage;
- paired H20, H800, and H100 comparisons against latest main's default path, a forced FP32 path, and #52696 if still relevant;
- separate reporting for score+TopK latency, full indexer latency, peak HBM, and end-to-end TTFT;
- real DeepSeek V3.2 prefill, MRCR/NIAH, and standard accuracy tasks.

We propose an implementation evidence floor of at least 1.10x indexer-stage geometric-mean speedup over the target validation matrix and at least 3% TTFT improvement in two long-context configurations.

#### Implementation plan

1. **PR 1: H64 / DeepSeek V3.2.** Production H64 kernel, static CUDA build, custom op, prefill dispatcher, tests, and benchmarks only.
2. **PR 2: H32 / GLM-5.1/5.2.** Reuse the reviewed interface and test contract, with separate tuning and validation.

Before submission, we will remove experimental selector branches, retain one production implementation, and add SPDX headers, MIT provenance, source commits, and a human-review record.

#### Core request for comments

We would like maintainer feedback on one decision: **would vLLM accept the H64 SM90 fused prefill indexer first as an opt-in experimental backend, with H32 in a separate PR after the interface and validation method are established?**

### Feedback Period.

one week.

### Any Other Things.

AI assistance was used for investigation, implementation experiments, and RFC drafting. The human submitter will review the RFC, code, and test results before requesting maintainer review.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by Heisenberg-Yin** [comment]
Hello, thank you for your interest in our LiteTopK work. We noticed that your approach appears to share several key design ideas with LiteTopK, including avoiding materialization of the full score tensor in HBM, using an on-chip score ring, and performing histogram/radix-based selection within the bucket containing the Top-K cutoff. These ideas appear very similar to those used in our LiteTopK PR #48726.

If you are interested, we would be happy to incorporate this LiteTopk-style SM90 implementation into our LiteTopK PR and collaborate with you on extending LiteTopK support to SM90, adding you as a contributor for PR #48726.

Best
Ziqi

**Comment by fxfxfxfxfxfxfxfx** [comment]
> Hello, thank you for your interest in our LiteTopK work. We noticed that your approach appears to share several key design ideas with LiteTopK, including avoiding materialization of the full score tensor in HBM, using an on-chip score ring, and performing histogram/radix-based selection within the bucket containing the Top-K cutoff. These ideas appear very similar to those used in our LiteTopK PR [#48726](https://github.com/vllm-project/vllm/pull/48726).
> 
> If you are interested, we would be happy to incorporate this LiteTopk-style SM90 implementation into our LiteTopK PR and collaborate with you on extending LiteTopK support to SM90, adding you as a contributor for PR [#48726](https://github.com/vllm-project/vllm/pull/48726).
> 
> Best Ziqi

Hi Ziqi,

Thank you for reaching out and for the generous collaboration proposal. We appreciate the work on LiteTopK, and we agree that the two projects share the same broad motivation: avoiding the cost of materializing the full score tensor in HBM and bringing score generation closer to TopK selection.

The two projects explore different ways of reaching that goal:

- LiteTopK reduces the full score tensor to a bounded candidate set and then performs selection over those candidates.
- Flash Indexer maintains the running exact TopK as scores are generated, so only the final indices are written to HBM.

These choices lead to different tradeoffs in intermediate storage and selection organization. The implementation discussed in LiteTopK currently targets SM100, while Flash Indexer focuses on SM90. We see them as independently developed, complementary implementations of a shared high-level idea.

We would be happy to collaborate on bringing the SM90 implementation into vLLM and to discuss the design choices in more detail. The implementation, benchmarks, and reports are available here:

https://github.com/fxfxfxfxfxfxfxfx/flash-indexer

For upstream integration, it may be useful to first align on a common interface, correctness contract, and benchmark methodology, while preserving the implementation history and contributions of both projects. We are open to whichever organization works best for review and long-term maintenance.

We would be glad to continue the discussion and work together on the integration.

Best,
Xiao Feng
